### PR TITLE
[INT-183] Use mkcert to generate and serve SSL for the custom server

### PR DESCRIPTION
**Pull request by archessmn**
Add a system requirement for `mkcert` when doing local SSL, update documentation and add yarn scripts to install and generate necessary certificates.

### CODE REVIEW

File: server/index.ts [hunk 1]
```diff
@@ -1,28 +1,41 @@
...
 import next from "next";
 import { Server, Socket } from "socket.io";
-import { DefaultEventsMap } from "socket.io/dist/typed-events";
 import { z } from "zod";
 import { authenticateSocket, isServerSocket } from "./auth";
 import { env, validateEnv } from "../lib/env.js";
+import { readFileSync } from "node:fs";
 
 const dev = env.NODE_ENV !== "production";
+const doSSL = env.DEV_SSL == "1"; // Used to decide whether or not to use https in a dev environment
```
Comment (markspolakovs) [inline server/index.ts:14]: We use `true` for `SLACK_ENABLED` but `1` here.

File: package.json [hunk 2]
```diff
@@ -4,11 +4,13 @@
...
   "scripts": {
     "dev": "nodemon",
-    "devSSL": "next dev --experimental-https",
+    "devSSL": "DEV_SSL=1 NODE_EXTRA_CA_CERTS=\"$(mkcert -CAROOT)/rootCA.pem\" nodemon",
     "build": "next build && tsc --project tsconfig.server.json",
     "start": "NODE_ENV=production node dist/server/index.js",
     "lint": "next lint",
     "test": "lint-staged",
+    "ssl:install": "mkcert -install",
+    "ssl:generate": "mkcert -key-file certificates/key.pem -cert-file certificates/cert.pem localhost 127.0.0.1 ::1",
```
Comment (markspolakovs) [inline package.json:13]: Could these be one command that does `mkcert -install && mkcert -generate`? Or perhaps more usefully, is there a reason they're not?

File: server/index.ts [hunk 3]
```diff
@@ -1,28 +1,41 @@
...
 
 validateEnv();
 
-app.prepare().then(() => {
-  const httpServer = createServer(handler);
+app.prepare().then(async () => {
+  let httpServer: HttpServer;
+
+  if (doSSL) {
+    const cert = await readFileSync(process.cwd() + "/certificates/cert.pem");
```
Comment (markspolakovs) [inline server/index.ts:32]: Should we add this to .gitignore?

**Review by markspolakovs — CHANGES_REQUESTED** [review]
Couple questions. Ideally we'd avoid a system dependency, but Filippo's mkcert is better than most things you'll find on NPM so it may be worth it.

File: server/index.ts (hunk 1 above)
Comment (markspolakovs) [reply server/index.ts:14]: Also that comment feels like it should be in `env.ts`.

File: package.json (hunk 2 above)
Comment (archessmn) [reply package.json:13]: Not sure why I did it that way actually

File: server/index.ts (hunk 3 above)
Comment (archessmn) [reply server/index.ts:32]: Don't know why but it already is boss

https://github.com/ystv/internal-site/blob/0b4403073e4bc1db6dad888dd234c2114f654cf9/.gitignore#L51

Comment (markspolakovs) [reply server/index.ts:32]: s🅱️icy

File: server/index.ts (hunk 1 above)
Comment (archessmn) [reply server/index.ts:14]: Agree with the comment. I did have a reasoning for it being 1 not true but realised something else I thought it relied on wasn't needed and so I've changed it.

File: server/index.ts [hunk 4]
```diff
@@ -1,28 +1,41 @@
...
 import next from "next";
 import { Server, Socket } from "socket.io";
-import { DefaultEventsMap } from "socket.io/dist/typed-events";
 import { z } from "zod";
 import { authenticateSocket, isServerSocket } from "./auth";
 import { env, validateEnv } from "../lib/env.js";
+import { readFileSync } from "node:fs";
 
 const dev = env.NODE_ENV !== "production";
+const doSSL = env.DEV_SSL == "true";
```
Comment (markspolakovs) [inline server/index.ts:14]: if we ran eslint in CI it'd shout at you about this `==` not being a `===`

**Review by markspolakovs — APPROVED** [review]
LGTM

**Comment by archessmn** [comment]
I would have preferred not to have a system dependency, but the mkcert binary packaged for npm doesn't actually work, and using the version that provides a typescript API doesn't allow you to trust the local root CA, so I think this is the best option currently.

**Pull request merged**